### PR TITLE
closes#1694 Configurable timezone for Rally logging

### LINKED ISSUE #1694: Configurable timezone for Rally logging
As discussed in https://discuss.elastic.co/t/how-to-set-other-time-zone-in-rallys-log/328683/2, right now the out of the box [logging.json](https://github.com/elastic/rally/blob/71688bcf5175cd1330481e8021f51d9c525785e1/esrally/resources/logging.json#L7) uses a [UTC formatter](https://github.com/elastic/rally/blob/71688bcf5175cd1330481e8021f51d9c525785e1/esrally/log.py#L29-L36) without a possibility to change it.

We could support configuring the timezone in a user friendly manner e.g. via an additional `timezone` property. A primitive, insecure and naive implementation could be as simple as:

```diff
$ git diff
diff --git a/esrally/log.py b/esrally/log.py
index 189a1b8..9423b1a 100644
--- a/esrally/log.py
+++ b/esrally/log.py
@@ -32,7 +32,10 @@ def configure_utc_formatter(*args, **kwargs):
     timestamps across all deployments regardless of machine settings.
     """
     formatter = logging.Formatter(fmt=kwargs["format"], datefmt=kwargs["datefmt"])
-    formatter.converter = time.gmtime
+    if user_tz := kwargs.get("timezone"):
+        formatter.converter = getattr(time, user_tz)
+    else:
+        formatter.converter = time.gmtime
     return formatter
 ```

which would allow changing the logging config file to e.g.:

```diff
diff --git a/esrally/resources/logging.json b/esrally/resources/logging.json
index ba559e2..91cca4c 100644
--- a/esrally/resources/logging.json
+++ b/esrally/resources/logging.json
@@ -4,6 +4,7 @@
     "normal": {
       "format": "%(asctime)s,%(msecs)d %(actorAddress)s/PID:%(process)d %(name)s %(levelname)s %(message)s",
       "datefmt": "%Y-%m-%d %H:%M:%S",
+      "timezone": "localtime",
       "()": "esrally.log.configure_utc_formatter"
     },
     "profile": {
```

### LINKED ISSUE #1700: `make prereq` is failing while setting up development environment
**Rally version** (get with `esrally --version`): 2.7.1

**Invoked command**: `make prereq`

**Configuration file (located in `~/.rally/rally.ini`))**:

**JVM version**: 

**OS version**: 

**Description of the problem including expected versus actual behavior**: Should install all the python version mentioned in `.ci/variables.json` however failing with this error:
```
pyenv install --skip-existing null
python-build: definition not found: null
```

Possibly because of a tag error for 3.11.2

**Steps to reproduce**:
 1. Run `make prereq`

**Provide logs (if relevant)**:

**Describe the feature**:

** comments**
After Typo fix 
```python
diff --git a/.ci/variables.json b/.ci/variables.json
--- a/.ci/variables.json
+++ b/.ci/variables.json
@@ -9,7 +9,7 @@
     "PY38": "3.8.16",
     "PY39": "3.9.16",
     "PY310": "3.10.10",
-    "PY310": "3.11.2",
+    "PY311": "3.11.2",
     "MIN_PY_VER": "3.8.13"
   }
 }
 ```
 The command might still fail with error 
 ```
 rm: .pyenv/shims/shims: is a directory
 make: *** [prereq] Error 1
 ```
 If it fails further, please take a look. Thanks for all your work!!!

### PR DESCRIPTION
**Pull request by vaibhavsolanki1193**
closes #1694, closes #1700
* Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)? YES
* Have you followed the [contributor guidelines](https://github.com/elastic/rally/blob/master/CONTRIBUTING.md)? YES
* Have you run `make check-all` successfully? YES
* Did you choose a [descriptive title and description](https://chris.beams.io/posts/git-commit/) for your PR? YES
* (Only for maintainers) Did you apply appropriate labels and a milestone? _SKIP_

### CODE REVIEW

**Comment by vaibhavsolanki1193** [comment]
I have signed the [Contributor Agreement](https://www.elastic.co/contributor-agreement). Please trigger the bot check again.

**Comment by pquentin** [comment]
@elasticmachine test this please

**Comment by pquentin** [comment]
buildkite build this please

**Comment by vaibhavsolanki1193** [comment]
@pquentin  please review once again. 
```
esrally race --track=http_logs --test-mode
```
is working now. Thank you

**Comment by pquentin** [comment]
buildkite build this please

**Comment by pquentin** [comment]
Thanks, I see all existing tests are passing now. I'll try to review this tomorrow.

**Review by pquentin — CHANGES_REQUESTED** [review]
Sorry for the delay. I finally took some time to dig into this. If we look at the proposed solutions, we have four of them:

1. This pull request: `datetime.utcnow().astimezone().tzinfo`
2. This pull request: `pytz.timezone()`
3. The GitHub issue: using `time.user_tz`
4. The Discuss topic: allow switching from `time.gmtime()` to `time.localtime()`

The problem is that most of them just don't work.

1. Is not reliable if you start Rally *during* a DST change (see PEP 495 for details). More importantly, it's also not reliable if Rally continues running after a timezone change: it will keep using the previous offset.
2. does not work in various timezones, see https://blog.ganssle.io/articles/2018/03/pytz-fastest-footgun.html for details (including why fixing it to use `localize()` would be slower)
3. is unreliable as `user_tz` ignores DST, so it would be off half of the year. There are ways to fix it (see https://docs.python.org/3/library/time.html#timezone-constants and the Note) but it would still suffer from the fact we won't update the UTC offset after a DST change.

While there are ways to fix 2. I think I showed that timezone are complicated enough that we should try to avoid them because there are probably other issues I'm not aware of. The fourth solution is the simplest and also the only that actually works.

Can you please implement the following changes?

* update your implementation so that setting `timezone` to `localtime` uses `time.localtime()` instead of `time.gmtime()` and remove all timezone considerations
* document this feature in https://esrally.readthedocs.io/en/stable/configuration.html#logging
* rename `configure_utc_formatter` to `configure_time_formatter`
* add tests for `configure_time_formatter` under `tests/log_test.py`

Then we'll be ready to merge!

Sorry that we're only noticing this now. Your work has been very valuable: having a working implementation made it easier to reason about possible issues.

**Comment by vaibhavsolanki1193** [comment]
Thank you @pquentin for the detailed inputs. I have modified the code as advised. Please review it once you get a chance!

**Comment by pquentin** [comment]
buildkite build this please

**Comment by pquentin** [comment]
Do you need help for adding the unit test?

**Comment by vaibhavsolanki1193** [comment]
The test cases were done however I was working around an issue I observed so please check the details below:

**Issue**: After this change gets merged, the local configuration file @ `~/.rally/logging.json` was still pointing to `esrally.log.configure_utc_formatter` which caused esrally fail to run. 
Running `./rally --help` gave this error:
```
Traceback (most recent call last):
  File "/home/codespace/.python/current/lib/python3.10/logging/config.py", line 392, in resolve
    self.importer(used)
ModuleNotFoundError: No module named 'esrally.log.configure_utc_formatter'; 'esrally.log' is not a package

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/codespace/.python/current/lib/python3.10/logging/config.py", line 544, in configure
    formatters[name] = self.configure_formatter(
  File "/home/codespace/.python/current/lib/python3.10/logging/config.py", line 656, in configure_formatter
    result = self.configure_custom(config)
  File "/home/codespace/.python/current/lib/python3.10/logging/config.py", line 471, in configure_custom
    c = self.resolve(c)
  File "/home/codespace/.python/current/lib/python3.10/logging/config.py", line 399, in resolve
    raise v
  File "/home/codespace/.python/current/lib/python3.10/logging/config.py", line 392, in resolve
    self.importer(used)
ValueError: Cannot resolve 'esrally.log.configure_utc_formatter': No module named 'esrally.log.configure_utc_formatter'; 'esrally.log' is not a package
```
**Steps tried to fix**:
```python
make prereq
make install
source .venv/bin/activate
./rally --help
```
I was expecting `make install` might also update the local logging.json, but that didn't happen. 

**Resolution**: Manually replaced `esrally.log.configure_utc_formatter` to `esrally.log.configure_time_formatter`

Please advise if you think this could be a breaking change or if handling the local logging.json is out of scope for esrally.

If this is considered an issue, a quick fix could be to replace the function name `configure_time_formatter` back to `configure_utc_formatter`. I would love to hear your thoughts on this. Thank you!

**Review by pquentin — CHANGES_REQUESTED** [review]
I'm sorry for the bad experience you're having here with the big delay between reviews. We're nearly there.

The main issue was the migration to the new formatter. To address this, I pushed https://github.com/elastic/rally/pull/1699/commits/2f547366ba17b6f4b87d25a22a8ec414314eb8b4 and https://github.com/elastic/rally/pull/1699/commits/f1dce9f4997bf7baf52d5dd5bc8a69ed99e58d2b. Please tell me what you think.

File: esrally/resources/logging.json [hunk 1]
```diff
@@ -4,12 +4,13 @@
     "normal": {
       "format": "%(asctime)s,%(msecs)d %(actorAddress)s/PID:%(process)d %(name)s %(levelname)s %(message)s",
       "datefmt": "%Y-%m-%d %H:%M:%S",
-      "()": "esrally.log.configure_utc_formatter"
+      "timezone": "localtime",
+      "()": "esrally.log.configure_time_formatter"
```
[the remark this answers is not in the document]
Comment (vaibhavsolanki1193) [reply esrally/resources/logging.json:8]: I also suggested the same but wanted to discuss the possibilities of other solutions. Thank you for your input and time.

**Comment by vaibhavsolanki1193** [comment]
no problem at all @pquentin. I also agree with the new changes.

**Comment by pquentin** [comment]
buildkite build this please

**Pull request closed without merging**